### ISSUE TITLE
BadDeviceId calling waveOutOpen

### ISSUE THREAD

**Issue by FutureInfinite** [naudio/NAudio#461]
I am using naudio to play audio files and is working fine. But I have a scenario where I am depositing an app on a Windows VM - and connected to it via MS RDP. I connect with audio being sent to system that the RDP is done one (i.e. client of VM). Audio will not play on client system and getting the
naudio BadDeviceId calling waveOutOpen
error in event viewer

Should it be possible to run this scenario and get audio playback on local system?

Note: I play the audio manually and comes across ok - so file is a good audio file

Peter

**Comment by markheath** [comment]
Yes, unfortunately WaveOut doesn't seem to work too well over RDP. What device id did you use? Did you try -1 (for wave mapper)

**Comment by FutureInfinite** [comment]
didn't specify any device ID - configured to operate with defaults.
doing the following:
--------------------------------------------------------------------------------------------------------
using (var audioFile = new AudioFileReader(fileName))
            {
                using (var outputDevice = new WaveOutEvent())
                {
                    outputDevice.Init(audioFile);
                    outputDevice.Play();
                }
            }
--------------------------------------------------------------------------------------------------------

Peter

**Comment by markheath** [comment]
Depending on the version of NAudio, the default is either 0, or -1. You can try setting them both explicitly, but as I've said, WaveOut doesn't seem to work well over RDP.

**Comment by MartinHolmes82** [comment]
What you recommend for playback over RDP?

**Comment by markheath** [comment]
I unfortunately don't know why NAudio doesn't work over RDP. It's not something I've ever needed so I've not investigated it. Maybe its to do with the WaveFormat - perhaps you could try a few different formats to see if that's the issue. You could also try DirectSoundOut - maybe that works better

**Comment by MartinHolmes82** [comment]
Cheers for the reply Mark. I've been developing an application for recording voice and sending for transcription. After seeing the comments I had a bit of a panic as all of our staff use RDP in ThinClients. But I tested it in the office and everything played back fine, just need to tweak a few bits as about half a second delay on playback to position bar.

**Comment by rahsharm77** [comment]
@MartinHolmes82  Can you please show an example snippet on how you managed to make this work. I am also getting the same error when I deploy my application to IIS.

My method to stream from a URL looks like:

        public static void PlayMp3FromUrl(Stream stream)
        {
            using (Stream ms = new MemoryStream())
            {
                using (stream)
                {
                    byte[] buffer = new byte[32768];
                    int read;
                    while ((read = stream.Read(buffer, 0, buffer.Length)) > 0)
                    {
                        ms.Write(buffer, 0, read);
                    }
                }

                ms.Position = 0;
                using (WaveStream blockAlignedStream =
                    new BlockAlignReductionStream(
                        WaveFormatConversionStream.CreatePcmStream(
                            new WaveFileReader(ms))))
                {
                    using (WaveOut waveOut = new WaveOut(WaveCallbackInfo.FunctionCallback()))
                    {
                        waveOut.Init(blockAlignedStream);
                        waveOut.DesiredLatency = 100; // 50ms latency
                        waveOut.Play();
                        while (waveOut.PlaybackState == PlaybackState.Playing)
                        {
                            System.Threading.Thread.Sleep(100);
                        }
                    }
                }
            }
        }

Any suggestions/comments would be greatly appreciated. Thanks in advance.

**Comment by deepakdubeydsys** [comment]
> @MartinHolmes82 Can you please show an example snippet on how you managed to make this work. I am also getting the same error when I deploy my application to IIS.
> 
> My method to stream from a URL looks like:
> 
> ```
>     public static void PlayMp3FromUrl(Stream stream)
>     {
>         using (Stream ms = new MemoryStream())
>         {
>             using (stream)
>             {
>                 byte[] buffer = new byte[32768];
>                 int read;
>                 while ((read = stream.Read(buffer, 0, buffer.Length)) > 0)
>                 {
>                     ms.Write(buffer, 0, read);
>                 }
>             }
> 
>             ms.Position = 0;
>             using (WaveStream blockAlignedStream =
>                 new BlockAlignReductionStream(
>                     WaveFormatConversionStream.CreatePcmStream(
>                         new WaveFileReader(ms))))
>             {
>                 using (WaveOut waveOut = new WaveOut(WaveCallbackInfo.FunctionCallback()))
>                 {
>                     waveOut.Init(blockAlignedStream);
>                     waveOut.DesiredLatency = 100;
>                     waveOut.Play();
>                     while (waveOut.PlaybackState == PlaybackState.Playing)
>                     {
>                         System.Threading.Thread.Sleep(100);
>                     }
>                 }
>             }
>         }
>     }
> ```
> 
> Any suggestions/comments would be greatly appreciated. Thanks in advance.

Hi @rahsharm77, 

We are getting the same problem. Could you please help us how you solve this ?

Thanks

**Comment by Summit-Information-Technologies** [comment]
> > @MartinHolmes82 Can you please show an example snippet on how you managed to make this work. I am also getting the same error when I deploy my application to IIS.
> > My method to stream from a URL looks like:
> > ```
> >     public static void PlayMp3FromUrl(Stream stream)
> >     {
> >         using (Stream ms = new MemoryStream())
> >         {
> >             using (stream)
> >             {
> >                 byte[] buffer = new byte[32768];
> >                 int read;
> >                 while ((read = stream.Read(buffer, 0, buffer.Length)) > 0)
> >                 {
> >                     ms.Write(buffer, 0, read);
> >                 }
> >             }
> > 
> >             ms.Position = 0;
> >             using (WaveStream blockAlignedStream =
> >                 new BlockAlignReductionStream(
> >                     WaveFormatConversionStream.CreatePcmStream(
> >                         new WaveFileReader(ms))))
> >             {
> >                 using (WaveOut waveOut = new WaveOut(WaveCallbackInfo.FunctionCallback()))
> >                 {
> >                     waveOut.Init(blockAlignedStream);
> >                     waveOut.DesiredLatency = 100;
> >                     waveOut.Play();
> >                     while (waveOut.PlaybackState == PlaybackState.Playing)
> >                     {
> >                         System.Threading.Thread.Sleep(100);
> >                     }
> >                 }
> >             }
> >         }
> >     }
> > ```
> > 
> > 
> > Any suggestions/comments would be greatly appreciated. Thanks in advance.
> 
> Hi @rahsharm77,
> 
> We are getting the same problem. Could you please help us how you solve this ?
> 
> Thanks

@deepakdubeydsys Is this still a problem or have you resolved it?

**Comment by rahsharm77** [comment]
> > @MartinHolmes82 Can you please show an example snippet on how you managed to make this work. I am also getting the same error when I deploy my application to IIS.
> > My method to stream from a URL looks like:
> > ```
> >     public static void PlayMp3FromUrl(Stream stream)
> >     {
> >         using (Stream ms = new MemoryStream())
> >         {
> >             using (stream)
> >             {
> >                 byte[] buffer = new byte[32768];
> >                 int read;
> >                 while ((read = stream.Read(buffer, 0, buffer.Length)) > 0)
> >                 {
> >                     ms.Write(buffer, 0, read);
> >                 }
> >             }
> > 
> >             ms.Position = 0;
> >             using (WaveStream blockAlignedStream =
> >                 new BlockAlignReductionStream(
> >                     WaveFormatConversionStream.CreatePcmStream(
> >                         new WaveFileReader(ms))))
> >             {
> >                 using (WaveOut waveOut = new WaveOut(WaveCallbackInfo.FunctionCallback()))
> >                 {
> >                     waveOut.Init(blockAlignedStream);
> >                     waveOut.DesiredLatency = 100;
> >                     waveOut.Play();
> >                     while (waveOut.PlaybackState == PlaybackState.Playing)
> >                     {
> >                         System.Threading.Thread.Sleep(100);
> >                     }
> >                 }
> >             }
> >         }
> >     }
> > ```
> > 
> > 
> > Any suggestions/comments would be greatly appreciated. Thanks in advance.
> 
> Hi @rahsharm77,
> 
> We are getting the same problem. Could you please help us how you solve this ?
> 
> Thanks

@deepakdubeydsys Is this still a problem at your end or have you resolved it?

**Comment by markheath** [comment]
Just following up on this as I'm closing out old issues with the help of AI. Here's a brief summary of the underlying issue and some recommendations...

`BadDeviceId calling waveOutOpen` is NAudio surfacing the Windows error `MMSYSERR_BADDEVICEID` returned by the operating system's `waveOutOpen` call — it is not an NAudio bug. It means Windows did not present a usable `waveOut` device for that session. This is a well-known limitation of the legacy `waveOut`/MME audio API in RDP sessions (and similarly in IIS / service /
session-0 / headless contexts): depending on the host and RDP audio-redirection configuration there may be no `waveOut` device enumerated for the session, so the call fails. That is exactly why another commenter on this thread found the same code worked fine over RDP from an in-office thin client — the outcome depends on the environment, not on NAudio.

Recommendations:

- For RDP / remote / server-hosted playback, do not use the legacy `WaveOut` backend. Use `WasapiOut` instead — it uses the modern Windows Core Audio (WASAPI) API and behaves much better with RDP and redirected audio.
- In current NAudio (v3) `WaveOut.DeviceNumber` already defaults to `-1` (the wave mapper), so the old "default device 0 vs -1" ambiguity no longer applies; switching to WASAPI is the real fix here.
- For IIS-hosted scenarios specifically: audio playback from a web-server process is generally not supported by Windows at all (no audio session) — that is an OS/hosting constraint unrelated to NAudio.

Since this is an OS / RDP environment limitation rather than an NAudio defect, and there has been no discussion for a few years, I will close this issue. Please reopen or comment if `WasapiOut` does not resolve it in your RDP setup and we can take another look.
